component/src/main/java/org/wso2/extension/siddhi/gpl/execution/streamingml/clustering/clustree/ClusTreeStreamProcessorExtension.java

```diff
@@ -0,0 +1,388 @@
+                        description = "This query will create a Siddhi app named ClusTreeTestSiddhiApp and will " +
+                                "accept 2D inputs od doubles. The query which is named query1 will create a ClusTree " +
+                                "model and will create a kmeans model after firsat 20 events and will " +
+                                "refresh it every 20 events after. Number of macro clusters will be 2 and the " +
+                                "maximum iterations of kmeans to converge will be 10. The max height of tree is " +
+                                "set to 5 so at maximum we will get 3^5 micro clusters from ClusTree and the " +
+                                "horizon is set as 50, so after 50 events micro clusters that were not updated " +
+                                "will lose their weight by half." //todo: add eg with default params
+                ),
+                @Example(
+                        syntax = "@App:name('ClusTreeTestSiddhiApp') \n" +
+                                "define stream InputStream (x double, y double);\n" +
+                                "@info(name = 'query1') \n" +
+                                "from InputStream#streamingml:ClusTree(2, x, y) \n" +
+                                "select closestCentroidCoordinate1, closestCentroidCoordinate2, x, y \n" +
+                                "insert into OutputStream;",
+                        description = "In this query one can note that the hyper parameters are not given. So they " +
+                                "will be set to their default values which are mentioned above. This mode of " +
+                                "querying is suggested for users who are not familier with ClusTree/KMeans algorithms."
+                )
+        }
+)
+public class ClusTreeStreamProcessorExtension extends StreamProcessor {
+    private final int minConstantParams = 1;
+    private final int maxConstantParams = 5;
+    private final int separateThreadThreshold = 500;
+    private int noOfClusters;
+    private int noOfEventsToRefreshMacroModel = 100; //todo : tune. check algo defaults
+    private int noOfDimensions;
+    private int maxIterations = 40;
+    private double[] coordinateValuesOfCurrentDataPoint;
+    private int noOfEventsReceived;
+    private ExecutorService executorService;
+    private ClusTreeModel clusTreeModel;
+    private KMeansModel kMeansModel;
+    private List<VariableExpressionExecutor> featureVariableExpressionExecutors = new LinkedList<>();
+    private static final Logger logger = Logger.getLogger(ClusTreeStreamProcessorExtension.class.getName());
+
+    @Override
+    protected List<Attribute> init(AbstractDefinition abstractDefinition, ExpressionExecutor[] expressionExecutors,
+                                   ConfigReader configReader, SiddhiAppContext siddhiAppContext) {
+        int noOfFeatures = inputDefinition.getAttributeList().size();
```

Comment:
this is maxNoOfFeatures, not # of features.

Reply:
fixed